### PR TITLE
Constrained random

### PR DESCRIPTION
**Pull request by cdonovick**
require https://github.com/leonardt/hwtypes/pull/23

### CODE REVIEW

File: fault/random.py [hunk 1]
```diff
@@ -1,10 +1,246 @@
...
+    def __iter__(self):
+        return iter(self._d)
+
+    def __len__(self):
+        return len(self._d)
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash(frozenset(self.items()))
+        return self._hash
```
Comment (leonardt) [inline fault/random.py]: It seems like the hash value for an instance is computed and fixed the first time `__hash__` is invoked, is this the intended behavior? What happens if we add items to the dictionary after hash has been called? Or is that a use case that we're not intending to support for this? If so, we should add some docs/guards against this (e.g. calling `__setitem__` after `_hash` has been set could raise an error).

Comment (leonardt) [reply fault/random.py]: Also, `__setitem__` is not defined, but looking at the rest of the code, it seems like this is used to construct a hashable version of a dictionary that isn't modified, so perhaps `FrozenDict` might be a better name?

File: fault/random.py [hunk 2]
```diff
@@ -1,10 +1,246 @@
...
+
+
+    def __call__(
+            self,
+            v_map : tp.Mapping[str, tp.Optional[int]],
+            pred : tp.Callable[..., AbstractBit],
+            N : int,
+            ) -> tp.AbstractSet[tp.Mapping[str, BitVector]]:
+        '''
+        V_map: map of labels to bitvector sizes, size is None indicates a Bit
```
Comment (leonardt) [inline fault/random.py]: Typo, `V_map` -> `v_map` to match lowercase parameter name

File: fault/random.py [hunk 3]
```diff
@@ -1,10 +1,246 @@
...
+        pred: function with signiture (v : AbstractBitVector[w] for v,w in V_map.items()) -> AbstractBit
```
Comment (leonardt) [inline fault/random.py]: Typo `signiture` -> `signature`

File: fault/random.py [hunk 4]
```diff
@@ -1,10 +1,246 @@
...
+                raise TypeError()
+        if solver.check() == z3.sat:
+            model = solver.model()
+            solver.pop()
+            return _model_to_hashdict(v_map, model)
+        else:
+            return None
+
+    def _compute_neighbors(self, v_map, pred, init, seen):
+        conditons = []
```
Comment (leonardt) [inline fault/random.py]: Typo `conditons` -> `conditions`

File: fault/random.py [hunk 5]
```diff
@@ -1,10 +1,246 @@
...
+                elif isinstance(v, AbstractBit):
+                    assignment[k] = random_bit()
+                else:
+                    TypeError()
+
+            assignment = HashableDict(assignment)
+            if assignment not in seen:
+                return assignment
+
+    def _find_closet(self, v_map, pred, seed):
```
Comment (leonardt) [inline fault/random.py]: Typo `_find_closet` -> `_find_closest` (I think?)

**Comment by cdonovick** [comment]
Your institution is correct.  I will add documentation and fix typos.

File: fault/random.py (hunk 1 above)
Comment (cdonovick) [reply fault/random.py]: __setitem__ is not defined to make the object immutable.  Changing the value of a hashable object will cause strange behavior as they will not be rehashed when they change.  Will change name to FrozenDict

**Pull request merged**